internal/filesystem/ipfs/files.go

```diff
@@ -0,0 +1,454 @@
+func (fe filesEntry) IsDir() bool  { return mfs.NodeType(fe.MfsLsEntry.Type) == mfs.TDir }
+func (fe filesEntry) Type() (mode fs.FileMode) {
+	if fe.IsDir() {
+		mode |= fs.ModeDir
+	}
+	return mode
+}
+
+func (fe filesEntry) Info() (fs.FileInfo, error) {
+	return &nodeInfo{
+		modTime: *fe.mountTime,
+		name:    fe.Name(),
+		size:    int64(fe.MfsLsEntry.Size),
+		mode:    fe.Type() | fe.permissions,
+	}, nil
+}
+
+func (ff *filesFile) Read(p []byte) (int, error) {
+	reader, err := ff.shell.FilesRead(
+		ff.ctx, ff.path,
+		shell.FilesRead.Count(int64(len(p))),
+		shell.FilesRead.Offset(ff.cursor),
+	)
+	if err != nil {
+		return -1, err
+	}
+	var errs []error
+	n, err := reader.Read(p)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if err := reader.Close(); err != nil {
+		errs = append(errs, err)
+	}
+	ff.cursor += int64(n)
+	return n, errors.Join(errs...)
+}
+
+func (ff *filesFile) Truncate(size int64) error {
+	return ff.root.Truncate(filesShellToFs(ff.path), size)
+}
+
+func (ff *filesFile) Write(p []byte) (int, error) {
+	err := ff.shell.FilesWrite(
+		ff.ctx, ff.path,
+		bytes.NewReader(p), shell.FilesWrite.Offset(ff.cursor),
+	)
+	// TODO: concern
+	// we don't know how many bytes were actually written
+	// Is there any way to get the API to tell us?
+	if err != nil {
+		return -1, err
+	}
+	written := len(p)
+	ff.cursor += int64(written)
+	return written, nil
+}
+
+func (ff *filesFile) Seek(offset int64, whence int) (int64, error) {
+	const op = "seek"
+	switch whence {
+	case io.SeekStart:
+		if offset < 0 {
+			err := generic.ConstError(
+				"tried to seek to a position before the beginning of the file",
+			)
+			return -1, newFSError(
+				op, filesShellToFs(ff.path),
+				err, fserrors.InvalidItem,
+			)
+		}
+		ff.cursor = offset
+	case io.SeekCurrent:
+		ff.cursor += offset
+	case io.SeekEnd:
+		info, err := ff.Stat()
+		if err != nil {
+			return -1, err
+		}
+		end := info.Size()
+		ff.cursor = end + offset
+	}
+	return ff.cursor, nil
+}
+
+func (f *filesFile) Close() error { f.cancel(); return nil }
+
+func filesClient(apiMaddr multiaddr.Multiaddr) (*shell.Shell, error) {
+	address, client, err := prepareClientTransport(apiMaddr)
+	if err != nil {
+		return nil, err
+	}
+	return shell.NewShellWithClient(address, client), nil
+}
+
+func prepareClientTransport(apiMaddr multiaddr.Multiaddr) (string, *http.Client, error) {
+	// TODO: magic number; decide on good timeout and const it.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	resolvedMaddr, err := resolveMaddr(ctx, apiMaddr)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// TODO: I think the upstream package needs a patch to handle this internally.
+	// we'll hack around it for now. Investigate later.
+	// (When trying to use a unix socket for the IPFS maddr
+	// the client returned from httpapi.NewAPI will complain on requests - forgot to copy the error lol)
```

Comment:
This (now) only applies to the new API, not the old one.
(This workaround was originally written sometime in 2018, and a fix was incorporated upstream sometime in 2020.)

I've filed an issue upstream along with a patch for the new API so that we can remove the workaround for both IPFS APIs.
(Dropping the workaround in this patchset can be done now though since we're using the old IPFS API that has the fix.)